File: tests/support/font_metrics.mjs
// Node-side font metrics mirror of src/font_metrics.rs.
//
// Both sides read the SAME DejaVu Sans / DejaVu Sans Bold TTFs. Same
// inputs → same outputs. This is the anchor that lets the reference
// pipeline and the Rust renderer agree byte-for-byte without running
// a shared JS runtime.
//
// API mirrors the Rust module:
//   charWidth(ch, family, size, bold)
//   textWidth(text, family, size, bold)
//   lineHeight(family, size)
//   ascent(family, size)
//   descent(family, size)
//
// Font family resolution: Java's logical "SansSerif" / "Dialog" and
// anything not explicitly monospaced → DejaVu Sans. "Monospaced" /
// "Courier" (Java logical name, no " New") / CSS "monospace" → DejaVu
// Sans Mono. mermaid's default font list starts with "trebuchet ms" —
// uninstalled on the reference machine → falls back to sans.

import opentype from 'opentype.js';
import { readFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';

const HERE = dirname(fileURLToPath(import.meta.url));
const FONTS_DIR = join(HERE, 'fonts');

function loadFace(fileName) {
  const buf = readFileSync(join(FONTS_DIR, fileName));
  const font = opentype.parse(buf.buffer.slice(buf.byteOffset, buf.byteOffset + buf.byteLength));
  // Extract the bits we actually need so we don't carry opentype
  // objects across hot paths.
  const upem = font.unitsPerEm;
  const os2 = font.tables.os2 ?? {};
  const hhea = font.tables.hhea ?? {};
  // DejaVu reports ascender/descender via hhea in design units.
  const ascender = hhea.ascender ?? os2.sTypoAscender ?? 0;
  const descender = hhea.descender ?? os2.sTypoDescender ?? 0; // negative
  const typoAscender = os2.sTypoAscender ?? ascender;
  return {
    upem,
    ascender,
    descender,
    typoAscender,
    glyphAdvance(codepoint) {
      const glyph = font.charToGlyph(String.fromCodePoint(codepoint));
      if (!glyph || glyph.index === 0) return null;
      return glyph.advanceWidth;
    },
  };
}

const FACES = {
  sans: loadFace('DejaVuSans.ttf'),
  sansBold: loadFace('DejaVuSans-Bold.ttf'),
};

function resolveFace(family, bold) {
  const primary = (family ?? '').split(',')[0].trim().toLowerCase();
  // Mono family not included yet — mermaid default never resolves to it.
  // If a future fixture forces us into mono, we'll add DejaVuSansMono.
  const _isMono =
    primary === 'monospaced' || primary === 'monospace' || primary === 'courier';
  // For MVP we treat mono as sans too (both Rust and here) — the sister
  // project bakes all four faces; we'll add the mono ones when a real
  // fixture needs them.
  return bold ? FACES.sansBold : FACES.sans;
}

export function charWidth(ch, family, size, bold = false) {
  if (ch === '\n' || ch === '\r') return 0;
  const face = resolveFace(family, bold);
  const cp = ch.codePointAt(0);
  const adv = face.glyphAdvance(cp);
  if (adv != null) return (adv / face.upem) * size;
  // Fallback: space advance
  const sp = face.glyphAdvance(0x20);
  if (sp != null) return (sp / face.upem) * size;
  return size * 0.6;
}

export function textWidth(text, family, size, bold = false) {
  let w = 0;
  for (const ch of text ?? '') w += charWidth(ch, family, size, bold);
  return w;
}

export function lineHeight(family, size) {
  const face = resolveFace(family, false);
  return ((face.ascender + Math.abs(face.descender)) / face.upem) * size;
}

export function ascent(family, size) {
  const face = resolveFace(family, false);
  return (face.ascender / face.upem) * size;
}

export function descent(family, size) {
  const face = resolveFace(family, false);
  return (Math.abs(face.descender) / face.upem) * size;
}

// Convenience: measure multi-line text exactly the way mermaid's
// getBBox shim needs it.
export function measureTextBlock(text, family, size, bold = false) {
  const lines = (text ?? '').split('\n');
  let maxW = 0;
  for (const l of lines) {
    const w = textWidth(l, family, size, bold);
    if (w > maxW) maxW = w;
  }
  const h = lineHeight(family, size) * lines.length;
  return { width: maxW, height: h };
}

File: src/font_metrics.rs
//! Font metrics computed from pre-extracted static DejaVu font data.
//!
//! Uses [`crate::font_data`] lookup tables instead of runtime TTF parsing.
//!
//! Vendored from the sister project
//! [plantuml-little](https://github.com/kookyleo/plantuml-little) at commit
//! `b32d6aa`, under its MIT-compatible multi-license. Mermaid has no runtime
//! Java dependency, but uses the same DejaVu TTF files, so the same glyph
//! advance math yields byte-exact geometry on this side of the pipeline too.

use crate::font_data::{FontMeta, DEJAVU_MONO, DEJAVU_MONO_BOLD, DEJAVU_SANS, DEJAVU_SANS_BOLD};

// ── Font family resolution ──────────────────────────────────────────────

/// Map a logical font family name to a canonical key.
/// Java logical fonts: "SansSerif"/"Dialog"→ DejaVu Sans, "Monospaced"/"Courier"→ DejaVu Sans Mono.
/// Physical fonts not installed on the reference machine (e.g. "Courier New", "Arial")
/// fall back to Dialog (sans-serif) in Java AWT.
/// For CSS `font-family` lists like "Courier New,monospace", we resolve based on
/// the PRIMARY (first) name — Java AWT uses the first name for font lookup.
fn resolve_face(family: &str, bold: bool) -> &'static FontMeta {
    // Use the first name in a CSS comma-separated font-family list
    let primary = family.split(',').next().unwrap_or(family).trim();
    let p = primary.to_lowercase();
    // Java logical font "Monospaced" and its alias "Courier" (without "New") map to mono.
    // CSS generic "monospace" also maps to mono.
    // "Courier New" is a physical font — uninstalled on reference machine → Dialog fallback.
    let is_mono = p == "monospaced" || p == "monospace" || p == "courier";
    if is_mono {
        if bold {
            &DEJAVU_MONO_BOLD
        } else {
            &DEJAVU_MONO
        }
    } else if bold {
        &DEJAVU_SANS_BOLD
    } else {
        &DEJAVU_SANS
    }
}

// ── Public API (signatures preserved from previous implementation) ───────

/// Width of a single character in the given font configuration.
///
/// Computes `glyph_hor_advance / units_per_em * size`, matching Java's
/// `font.getStringBounds(ch, frc).getWidth()` with `FRACTIONALMETRICS_ON`.
pub fn char_width(ch: char, family: &str, size: f64, bold: bool, _italic: bool) -> f64 {
    if ch == '\n' || ch == '\r' {
        return 0.0;
    }
    let face = resolve_face(family, bold);
    let upem = face.units_per_em as f64;
    if let Some(adv) = face.glyph_advance(ch as u32) {
        return adv as f64 / upem * size;
    }
    // Fallback: use space advance for unmapped characters
    if let Some(sp_adv) = face.glyph_advance(' ' as u32) {
        return sp_adv as f64 / upem * size;
    }
    size * 0.6 // last-resort fallback
}

/// Total width of a text string (sum of character advances).
pub fn text_width(text: &str, family: &str, size: f64, bold: bool, italic: bool) -> f64 {
    text.chars()
        .map(|c| char_width(c, family, size, bold, italic))
        .sum()
}

/// Line height = ascent + |descent| (leading is 0 for DejaVu fonts).
///
/// Matches Java's `LineMetrics.getHeight()`.
pub fn line_height(family: &str, size: f64, _bold: bool, _italic: bool) -> f64 {
    let face = resolve_face(family, false); // vertical metrics are style-independent
    let upem = face.units_per_em as f64;
    let asc = face.ascender as f64; // positive (hhea.ascender)
    let desc = face.descender.unsigned_abs() as f64; // make positive
    (asc + desc) / upem * size
}

/// Font ascent (baseline to top of tallest glyph).
///
/// Matches Java's `LineMetrics.getAscent()`.
pub fn ascent(family: &str, size: f64, _bold: bool, _italic: bool) -> f64 {
    let face = resolve_face(family, false);
    face.ascender as f64 / face.units_per_em as f64 * size
}

/// Font descent (baseline to bottom of lowest glyph).
///
/// Matches Java's `LineMetrics.getDescent()` (positive value).
pub fn descent(family: &str, size: f64, _bold: bool, _italic: bool) -> f64 {
    let face = resolve_face(family, false);
    face.descender.unsigned_abs() as f64 / face.units_per_em as f64 * size
}

/// OS/2 typographic ascent. Used for DOT cluster label dimensions which match
/// Java's `StringBounder.calculateDimension()` text block height.
pub fn typo_ascent(family: &str, size: f64, _bold: bool, _italic: bool) -> f64 {
    let face = resolve_face(family, false);
    let upem = face.units_per_em as f64;
    let typo_asc = face.typo_ascender as f64;
    typo_asc / upem * size
}

// ── Tests ───────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    // Java ground truth (FRACTIONALMETRICS_ON, DejaVu Sans):
    // SansSerif 12 PLAIN: ascent=11.1386718750 descent=2.8300781250 height=13.9687500000
    // SansSerif 13 PLAIN: ascent=12.0668945313 descent=3.0659179688 height=15.1328125000
    // SansSerif 18 PLAIN: ascent=16.7080078125 descent=4.2451171875 height=20.9531250000
    // charW('W') at 12 = 11.8652343750
    // width('foo1') at 12 = 26.5429687500

    #[test]
    fn ascent_matches_java() {
        let a12 = ascent("SansSerif", 12.0, false, false);
        let a13 = ascent("SansSerif", 13.0, false, false);
        let a18 = ascent("SansSerif", 18.0, false, false);
        assert!((a12 - 11.1386718750).abs() < 1e-6, "a12={a12}");
        assert!((a13 - 12.0668945313).abs() < 1e-6, "a13={a13}");
        assert!((a18 - 16.7080078125).abs() < 1e-6, "a18={a18}");
    }

    #[test]
    fn descent_matches_java() {
        let d12 = descent("SansSerif", 12.0, false, false);
        assert!((d12 - 2.8300781250).abs() < 1e-6, "d12={d12}");
    }

    #[test]
    fn line_height_matches_java() {
        let h12 = line_height("SansSerif", 12.0, false, false);
        let h13 = line_height("SansSerif", 13.0, false, false);
        let h18 = line_height("SansSerif", 18.0, false, false);
        assert!((h12 - 13.9687500000).abs() < 1e-6, "h12={h12}");
        assert!((h13 - 15.1328125000).abs() < 1e-6, "h13={h13}");
        assert!((h18 - 20.9531250000).abs() < 1e-6, "h18={h18}");
    }

    #[test]
    fn char_width_w_matches_java() {
        let w = char_width('W', "SansSerif", 12.0, false, false);
        assert!((w - 11.8652343750).abs() < 1e-6, "W width={w}");
    }

    #[test]
    fn text_width_foo1_matches_java() {
        let w = text_width("foo1", "SansSerif", 12.0, false, false);
        assert!((w - 26.5429687500).abs() < 1e-4, "foo1 width={w}");
    }

    #[test]
    fn monospaced_metrics() {
        // All monospaced chars should have equal advance width
        let w_a = char_width('a', "Monospaced", 13.0, false, false);
        let w_w = char_width('W', "Monospaced", 13.0, false, false);
        assert!(
            (w_a - w_w).abs() < 1e-6,
            "mono: a={w_a} W={w_w} should be equal"
        );
    }

    #[test]
    fn bold_width_differs() {
        let w_plain = char_width('W', "SansSerif", 12.0, false, false);
        let w_bold = char_width('W', "SansSerif", 12.0, true, false);
        assert!(w_bold > w_plain, "bold W should be wider");
    }

    #[test]
    fn family_resolution() {
        // "Monospaced" (Java logical name) resolves to mono font
        let w_mono = char_width('a', "Monospaced", 12.0, false, false);
        let w_monospace = char_width('a', "monospace", 12.0, false, false);
        assert!((w_mono - w_monospace).abs() < 1e-10);
        // "Courier" (Java logical font, no "New") maps to Monospaced
        let w_courier = char_width('a', "Courier", 12.0, false, false);
        assert!((w_mono - w_courier).abs() < 1e-10, "Courier maps to mono");
        // "Courier New" is a physical font not installed on reference machine
        // → Java Dialog fallback → sans-serif
        let w_courier_new = char_width('a', "Courier New", 12.0, false, false);
        let w_sans = char_width('a', "SansSerif", 12.0, false, false);
        assert!(
            (w_courier_new - w_sans).abs() < 1e-10,
            "Courier New maps to sans (Dialog fallback)"
        );
        // "SansSerif", "Dialog", "Arial" all resolve to sans font
        let w3 = char_width('a', "SansSerif", 12.0, false, false);
        let w4 = char_width('a', "Dialog", 12.0, false, false);
        assert!((w3 - w4).abs() < 1e-10);
    }

    #[test]
    fn arbitrary_size_works() {
        // Size 15 was not in the old lookup table — runtime computation handles any size
        let h = line_height("SansSerif", 15.0, false, false);
        assert!(h > 0.0);
        assert!((h - (1901.0 + 483.0) / 2048.0 * 15.0).abs() < 1e-6);
    }

    #[test]
    fn text_width_matches_java_reference() {
        // Verify text_width matches Java PlantUML's getStringBounds for various strings
        let cases: &[(&str, f64, bool, f64)] = &[
            ("Alice", 14.0, false, 33.667),
            ("Bob", 14.0, false, 27.0566),
            ("Hello", 13.0, false, 32.9507),
            ("Test", 14.0, false, 29.9482),
            ("Grouping messages", 13.0, true, 144.5869),
            ("Swimlane1", 18.0, false, 98.6484),
            ("Action 1", 12.0, false, 49.2422),
        ];
        for (text, size, bold, java_w) in cases {
            let our_w = text_width(text, "SansSerif", *size, *bold, false);
            assert!(
                (our_w - java_w).abs() < 0.001,
                "text_width(\"{text}\", size={size}, bold={bold}): ours={our_w:.4}, java={java_w:.4}"
            );
        }
    }
}
